Drive_append_to_sheet: Append columns to a Google Sheet

mulbc opened the pull request:
##### SUMMARY
This module allow appending data to a Google Sheet.
This can be useful if you want to gather lots of data from
different amounts of hosts and process it further.

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
drive_append_to_sheet

##### ADDITIONAL INFORMATION
Using this module it is possible to archive variables in a playbook run into a Google Sheet.
This is especially interesting if you are trying to gather facts across lots of servers and you will want to do some statistical analysis on that data later on.
The idea for this module was started when I realized that just writing to a local file from different hosts in parallel (with delegate_to: localhost) does not work atomically.
#30413
#47322

Commit:
Drive_append_to_sheet: Append columns to a Google Sheet

This module allow appending data to a Google Sheet.
This can be useful if you want to gather lots of data from
different amounts of hosts and process it further.

Signed-off-by: Chris Blum <chris.n.blum@gmail.com>

dagwieers commented on lib/ansible/modules/cloud/google/drive_append_to_sheet.py:
```diff
@@ -0,0 +1,203 @@
+      you will want to do some statistical analysis on that data later on.
+      A Google account and the Python client library for Google APIs are requirements.
+      Also you must have the Google Sheets API activated and valid credentials for this handy.
+      The necessary API steps can be done here:
```
Use a `seealso:` sections for external information.

mulbc replied:
resolved by fb49811

dagwieers commented on lib/ansible/modules/cloud/google/drive_append_to_sheet.py:
```diff
@@ -0,0 +1,203 @@
+    - "Using this module it is possible to archive variables in a playbook run into a Google Sheet.
+      This is especially interesting if you are trying to gather facts across lots of servers and
+      you will want to do some statistical analysis on that data later on.
+      A Google account and the Python client library for Google APIs are requirements.
```
The python library should go into `requirements:`. The Google account is actually related to a parameter, so that information should go with the required parameter.

mulbc replied:
I think what you meant was already in there - but I removed it from the description...

dagwieers commented on lib/ansible/modules/cloud/google/drive_append_to_sheet.py:
```diff
@@ -0,0 +1,203 @@
+
+description:
+    - "Using this module it is possible to archive variables in a playbook run into a Google Sheet.
+      This is especially interesting if you are trying to gather facts across lots of servers and
```
Please use different items in the description list for different paragraphs.

mulbc replied:
resolved by fb49811

dagwieers commented on lib/ansible/modules/cloud/google/drive_append_to_sheet.py:
```diff
@@ -0,0 +1,203 @@
+---
+module: drive_append_to_sheet
+
+short_description: Archive variables in a Google Sheet
```
The module can be used for more than archiving (which is just a use-case).
So I would make this more generic.

mulbc replied:
resolved by fb49811

dagwieers commented on lib/ansible/modules/cloud/google/drive_append_to_sheet.py:
```diff
@@ -0,0 +1,203 @@
+      https://developers.google.com/sheets/api/quickstart/python#step_1_turn_on_the"
+
+options:
+    columns:
```
The parameter is plural, while the description is singular.
Also, please add type for each parameter so this is added properly to the documentation.

mulbc replied:
resolved by fb49811

dagwieers commented on lib/ansible/modules/cloud/google/drive_append_to_sheet.py:
```diff
@@ -0,0 +1,203 @@
+    sheetName:
+        description:
+            - This is the Sheet inside of the GSheet that we will add information
+        required: false
```
If it is not required, don't add it.

mulbc replied:
I disagree - since sometimes you would want to select which Sheet inside of the GSheet (I didn't invent that confusing naming) you would want that information to be in. If the field isn't set - it will fill the information into the first sheet - which should be an ok default.

dagwieers commented on lib/ansible/modules/cloud/google/drive_append_to_sheet.py:
```diff
@@ -0,0 +1,203 @@
+        aliases: [ 'name' ]
+    sheetID:
+        description:
+            - This is the Sheet ID
```
Please don't start every description with "This is", simply state what it is used for.
```suggestion
            - The ID of sheet to append columns to.
```

dagwieers commented on lib/ansible/modules/cloud/google/drive_append_to_sheet.py:
```diff
@@ -0,0 +1,203 @@
+    client_id:
+        description:
+            - This is the Google OAuth client_id
+              It ends with .apps.googleusercontent.com
```
Make this two items in your description.
```suggestion
            - The value should end with C(.apps.googleusercontent.com).
```

mulbc replied:
resolved by fb49811

dagwieers commented on lib/ansible/modules/cloud/google/drive_append_to_sheet.py:
```diff
@@ -0,0 +1,203 @@
+'''
+
+EXAMPLES = '''
+# Append one row with three columns
```
This should be your name-value.
```suggestion
- name: Append one row with three columns
```

dagwieers commented on lib/ansible/modules/cloud/google/drive_append_to_sheet.py:
```diff
@@ -0,0 +1,203 @@
+
+EXAMPLES = '''
+# Append one row with three columns
+- name: Write stuff to GSheet
```
This is useless.
```suggestion
```

dagwieers commented on lib/ansible/modules/cloud/google/drive_append_to_sheet.py:
```diff
@@ -0,0 +1,203 @@
+    description: The columns that were appended
+    returned: success
+    type: list
+message:
```
The standard parameter is named `msg:`, so it is better to keep it consistent with all other modules.

mulbc replied:
resolved by fb49811

dagwieers commented on lib/ansible/modules/cloud/google/drive_append_to_sheet.py:
```diff
@@ -0,0 +1,203 @@
+    # define available arguments/parameters a user can pass to the module
+    module_args = dict(
+        columns=dict(type='list', required=True, aliases=['name']),
+        sheetID=dict(required=True),
```
Please add types to every parameter in the argspec.
So if you did not add it, it most likely should be: `type='str'`.

mulbc replied:
resolved by fb49811

dagwieers reviewed (changes requested):
Apart from the cosmetic/doc changes, I have some other remarks:
- I don't think it belongs in cloud, because that's about managing cloud infrastructure
- I wonder if it makes sense to have this kind of highly specialized module in Ansible. I am torn between making one module that is a lot more flexible in what it can do, or a set of modules that offers a complete interface. But this module is very specific/limited.

Commit:
Implement review comments

Implemented review comments - This consisted mostly
of changes in the documentation, but also
changed the return label from `message` to `msg`

Signed-off-by: Chris Blum <chris.n.blum@gmail.com>

rambleraptor commented:
Hi! I'm one of the engineers working on the GCP modules.

I agree that this probably doesn't belong in the cloud/ folder. I'm not entirely sure where it would belong though. 

Any core members have any ideas?

dagwieers committed:
Update lib/ansible/modules/cloud/google/drive_append_to_sheet.py

Implement review suggestions

Co-Authored-By: mulbc <zeichenanonym@web.de>

mulbc commented:
I had a brief discussion with @dagwieers in IRC yesterday and since there isn't anything for the Google Drive API yet and this might create more than one module (and thus might get new folders in the module hierarchy) he created a wiki page for this:
https://github.com/ansible/community/wiki/GDrive
Current goal is to look at which API methods might be interesting as an Ansible module and plan ahead how the general interface of these future Google Drive modules should look like - then proceed here.
